# Subsets

## Import Pandas

Aliasing pandas as `pd` is a widely adopted convention that simplifies the syntax for accessing its functionalities.\
After this statement, you can use `pd` to access all the functionalities provided by the pandas library.

In [1]:
# This line imports the pandas library and aliases it as 'pd'.

import pandas as pd

___

## Creating a `DataFrame` from a CSV file

The `pd.read_csv()` function is used to read the data from the file 'data/titanic.csv'.

In [2]:
# Load the Titanic dataset from a CSV file into a DataFrame named 'titanic'.

titanic = pd.read_csv('data/titanic.csv')

In [3]:
# Display the DataFrame 'titanic'.

titanic

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

___

## Selecting specific columns

![Selecting specific columns from a DataFrame](images/03_subset_columns.svg)

Access the 'Age' column from the `DataFrame` 'titanic' to return a `Series` object containing all the data in the 'Age' column:

In [4]:
titanic['Age']

0      22.0
1      38.0
2      26.0
3      35.0
4      35.0
       ... 
886    27.0
887    19.0
888     NaN
889    26.0
890    32.0
Name: Age, Length: 891, dtype: float64

In [5]:
# Check the type of the 'Age' column in 'titanic' using the 'type()' function.

type(titanic['Age'])

pandas.core.series.Series

Use the `shape` attribute to determine the dimensions of the `Series`.\
It returns a tuple representing the number of rows and columns (rows, columns).

In [6]:
titanic['Age'].shape

(891,)

___

## Calling multiple `Series`

Select columns 'Age' and 'Sex' from the 'titanic' `DataFrame` using double square brackets:

In [7]:
# Select the columns 'Age' and 'Sex' from the 'titanic' DataFrame.

titanic[['Age', 'Sex']]

Age     Sex
0    22.0    male
1    38.0  female
2    26.0  female
3    35.0  female
4    35.0    male
..    ...     ...
886  27.0    male
887  19.0  female
888   NaN  female
889  26.0    male
890  32.0    male

[891 rows x 2 columns]

```{note}
The inner square brackets define a Python `list` with column names, whereas the outer square brackets are used to select the data from a pandas `DataFrame` as seen in the previous example.

Select columns 'Age' and 'Sex' from the 'titanic' `DataFrame` using double square brackets, and then apply the `type()` function to the resulting `DataFrame` subset:

In [8]:
# Use the 'type()' function to determine the data type of the DataFrame subset.

type(titanic[['Age', 'Sex']])

pandas.core.frame.DataFrame

Select columns 'Age' and 'Sex' from the 'titanic' `DataFrame` using double square brackets, and then apply the `shape` attribute to the resulting `DataFrame` subset:

In [9]:
# Use the 'shape' attribute to determine the dimensions of the DataFrame subset.

titanic[['Age', 'Sex']].shape

(891, 2)

___

## Filtering rows based on conditional expressions

![Filtering specific rows from a DataFrame](images/03_subset_rows.svg)

We can use filtering to extract subsets of a `DataFrame`.

To select rows based on a conditional expression, use a condition inside the selection brackets `[]`.\
The condition inside the selection brackets checks for rows where the specified column's values exceed a certain threshold.

The code below returns a subset of the `DataFrame` containing only the rows where the 'Age' column has a value greater than 35.

In [10]:
# Filter rows in the 'titanic' DataFrame where the 'Age' column is greater than 35.

titanic[titanic['Age'] > 35]

PassengerId  Survived  Pclass  \
1              2         1       1   
6              7         0       1   
11            12         1       1   
13            14         0       3   
15            16         1       2   
..           ...       ...     ...   
865          866         1       2   
871          872         1       1   
873          874         0       3   
879          880         1       1   
885          886         0       3   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
13                         Andersson, Mr. Anders Johan    male  39.0      1   
15                    Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
..                                                 ...     ...   ...    ...   
865                           Bystrom, Mrs. (Karolina)  female  42.0      0   
871   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female  47.0      1   
873                        Vander Cruyssen, Mr. Victor    male  47.0      0   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
885               Rice, Mrs. William (Margaret Norton)  female  39.0      0   

     Parch    Ticket     Fare Cabin Embarked  
1        0  PC 17599  71.2833   C85        C  
6        0     17463  51.8625   E46        S  
11       0    113783  26.5500  C103        S  
13       5    347082  31.2750   NaN        S  
15       0    248706  16.0000   NaN        S  
..     ...       ...      ...   ...      ...  
865      0    236852  13.0000   NaN        S  
871      1     11751  52.5542   D35        S  
873      0    345765   9.0000   NaN        S  
879      1     11767  83.1583   C50        C  
885      5    382652  29.1250   NaN        Q  

[217 rows x 12 columns]

What happens inside the square brackets is that Pandas evaluates the condition 'titanic['Age'] > 35' element-wise, resulting in a boolean `Series` where `True` indicates that the corresponding passenger's age is greater than 35.

This is called a boolean mask.

In [11]:
# Create a boolean mask to filter rows where the age of passengers is greater than 35.

titanic['Age'] > 35

0      False
1       True
2      False
3      False
4      False
       ...  
886    False
887    False
888    False
889    False
890    False
Name: Age, Length: 891, dtype: bool

We can also use the `isin()` method to create a boolean mask, where `True` indicates that the corresponding 'Pclass' value is present in the specified list '[2, 3]'.

In [12]:
# Filter rows in the 'titanic' DataFrame where the 'Pclass' column values are either 2 or 3.

titanic[titanic['Pclass'].isin([2, 3])]

PassengerId  Survived  Pclass                                      Name  \
0              1         0       3                   Braund, Mr. Owen Harris   
2              3         1       3                    Heikkinen, Miss. Laina   
4              5         0       3                  Allen, Mr. William Henry   
5              6         0       3                          Moran, Mr. James   
7              8         0       3            Palsson, Master. Gosta Leonard   
..           ...       ...     ...                                       ...   
884          885         0       3                    Sutehall, Mr. Henry Jr   
885          886         0       3      Rice, Mrs. William (Margaret Norton)   
886          887         0       2                     Montvila, Rev. Juozas   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
0      male  22.0      1      0         A/5 21171   7.2500   NaN        S  
2    female  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S  
4      male  35.0      0      0            373450   8.0500   NaN        S  
5      male   NaN      0      0            330877   8.4583   NaN        Q  
7      male   2.0      3      1            349909  21.0750   NaN        S  
..      ...   ...    ...    ...               ...      ...   ...      ...  
884    male  25.0      0      0   SOTON/OQ 392076   7.0500   NaN        S  
885  female  39.0      0      5            382652  29.1250   NaN        Q  
886    male  27.0      0      0            211536  13.0000   NaN        S  
888  female   NaN      1      2        W./C. 6607  23.4500   NaN        S  
890    male  32.0      0      0            370376   7.7500   NaN        Q  

[675 rows x 12 columns]

```{note}

The above is equivalent to filtering by rows for which the class is either 2 or 3 and combining the two statements with an `|` (OR) operator.

When combining multiple conditional statements, each condition must be surrounded by parentheses `()`.\
Moreover, you can not use `or`/`and` but need to use the OR operator `|` and the AND operator `&`.

The following code uses boolean indexing with logical OR (`|`) to create a mask, where `True` indicates that the corresponding 'Pclass' value is either 2 or 3.

In [13]:
# Filter rows in the 'titanic' DataFrame where the 'Pclass' column values are either 2 or 3.

titanic[(titanic['Pclass'] == 2) | (titanic['Pclass'] == 3)]

PassengerId  Survived  Pclass                                      Name  \
0              1         0       3                   Braund, Mr. Owen Harris   
2              3         1       3                    Heikkinen, Miss. Laina   
4              5         0       3                  Allen, Mr. William Henry   
5              6         0       3                          Moran, Mr. James   
7              8         0       3            Palsson, Master. Gosta Leonard   
..           ...       ...     ...                                       ...   
884          885         0       3                    Sutehall, Mr. Henry Jr   
885          886         0       3      Rice, Mrs. William (Margaret Norton)   
886          887         0       2                     Montvila, Rev. Juozas   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
0      male  22.0      1      0         A/5 21171   7.2500   NaN        S  
2    female  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S  
4      male  35.0      0      0            373450   8.0500   NaN        S  
5      male   NaN      0      0            330877   8.4583   NaN        Q  
7      male   2.0      3      1            349909  21.0750   NaN        S  
..      ...   ...    ...    ...               ...      ...   ...      ...  
884    male  25.0      0      0   SOTON/OQ 392076   7.0500   NaN        S  
885  female  39.0      0      5            382652  29.1250   NaN        Q  
886    male  27.0      0      0            211536  13.0000   NaN        S  
888  female   NaN      1      2        W./C. 6607  23.4500   NaN        S  
890    male  32.0      0      0            370376   7.7500   NaN        Q  

[675 rows x 12 columns]

Remember that the `notna()` conditional function returns a `True` for each row where the values are not a 'null' value (`NaN`).\
As such, this can be combined with the selection brackets `[]` to filter the data table.

Filter rows in the 'titanic' `DataFrame` where 'Embarked' values are not null (not `NaN`).\
`notna()` returns `True` for non-null values, allowing us to select rows with valid embarkation data.

In [14]:
titanic[titanic['Embarked'].notna()]

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[889 rows x 12 columns]

We can compare the number of extracted rows with the number from the original `DataFrame`:

In [15]:
titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


Comparing the two number of rows (889 and 891) tells us, that two rows are missing embarkation data.

___

## Selecting rows and columns

![Selecting specific rows and columns from a DataFrame](images/03_subset_columns_rows.svg)

The following code uses boolean indexing to first filter rows where the age is greater than 35, and then selects specific columns 'Name' and 'Pclass' using double square brackets.

In [16]:
# Filter rows in the 'titanic' DataFrame where the age is greater than 35,
# then select only the 'Name' and 'Pclass' columns for these filtered rows.

titanic[titanic['Age'] > 35][['Name', 'Pclass']]

Name  Pclass
1    Cumings, Mrs. John Bradley (Florence Briggs Th...       1
6                              McCarthy, Mr. Timothy J       1
11                            Bonnell, Miss. Elizabeth       1
13                         Andersson, Mr. Anders Johan       3
15                    Hewlett, Mrs. (Mary D Kingcome)        2
..                                                 ...     ...
865                           Bystrom, Mrs. (Karolina)       2
871   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)       1
873                        Vander Cruyssen, Mr. Victor       3
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)       1
885               Rice, Mrs. William (Margaret Norton)       3

[217 rows x 2 columns]

```{admonition} A note on square bracket indexing []
Square bracket indexing (`[]`) is a versatile method for accessing data in Pandas DataFrames, but there are certain tasks that it cannot perform as efficiently or directly compared to `loc` and `iloc`. Here are a few limitations of square bracket indexing in comparison to `loc` and `iloc`:

1. **Positional Selection**: When the selection is purely positional (selecting the first five rows, or columns 2 to 4), `iloc` is the most straightforward tool for the job.

2. **Inclusive Slicing:** Square bracket slicing excludes the end point, requiring adjustment of slice endpoints. Conversely, `loc` allows for inclusive label-based slicing, simplifying the specification of row and column ranges.

3. **Dealing with Non-Integer Labels:** Square brackets may falter with non-integer labels or custom indices due to potential ambiguities, particularly when index labels might be confused with column names. `loc` ensures robust label-based selection, irrespective of label data type.

4. **Efficiency in Complex Selections:** For large datasets or intricate selection tasks, `loc` and `iloc` may provide enhanced performance due to their optimizations for label and integer indexing, respectively. They offer a more specialized approach for accessing and manipulating `DataFrame` data, particularly in complex scenarios.

We have attached a separate notebook that introduces `loc` and `iloc`, two properties of pandas `DataFrame` and `Series` objects that provide methods for indexing. 
```

___

## Key points

* When selecting subsets of data, square brackets `[]` are used.

* Inside these square brackets, you can use
    
    * a single column/row label
    
    * a list of column/row labels
    
    * a slice of labels
    
    * a conditional expression